File: backend/lumen_schemas/image_generation.py
"""Pydantic schemas for /api/v1/image-generation.

Spec: §4.3
"""
from datetime import datetime
from typing import Optional, Dict, Any
from pydantic import BaseModel, ConfigDict, Field


class ImageGenerationCreate(BaseModel):
    # model_config_id 触发 Pydantic v2 的 "model_" 保护命名空间警告;
    # protected_namespaces=() 显式禁用,让字段名带 "model_" 前缀。
    model_config = ConfigDict(protected_namespaces=())

    model_config_id: int
    prompt: str = Field(min_length=1, max_length=4000)
    negative_prompt: Optional[str] = Field(default=None, max_length=4000)
    size: str = "1024x1024"
    n: int = Field(default=1, ge=1, le=4)
    quality: Optional[str] = None
    style: Optional[str] = None
    extra_params: Optional[Dict[str, Any]] = None
    # M35: optional playbook id to inject style keywords into the prompt
    playbook_id: Optional[int] = None


class ImageGenerationListItem(BaseModel):
    # model_name / model_type 触发 Pydantic v2 "model_" 保护命名空间警告;
    # protected_namespaces=() 禁用掉。
    model_config = ConfigDict(protected_namespaces=())

    id: int
    prompt_preview: str
    model_config_id: int
    model_name: str
    model_type: str
    size: str
    status: str
    has_thumbnail: bool
    file_size: Optional[int]
    width: Optional[int]
    height: Optional[int]
    duration_ms: Optional[int]
    created_at: datetime


class ImageGenerationDetail(ImageGenerationListItem):
    prompt: str
    negative_prompt: Optional[str]
    quality: Optional[str]
    style: Optional[str]
    n: int
    params: Optional[Dict[str, Any]]
    error_message: Optional[str]
    updated_at: datetime

File: frontend/types/image-generation.ts
// frontend/types/image-generation.ts
// M22 — image generation feature (T14)
// Spec: docs/superpowers/specs/2026-06-11-image-generation-design.md §4.3
// Backend envelopes (PaginatedResponse / SingleResponse) live in
// app/schemas/common.py; the frontend reads them via `res.data.data` then
// `res.data.total/page/page_size` (see CLAUDE.md §3).

export type ImageGenStatus = "pending" | "generating" | "completed" | "failed";

/** Subset returned by GET /image-generation/ (list endpoint). */
export interface ImageGenerationListItem {
  id: number;
  prompt_preview: string;
  model_config_id: number;
  model_name: string;
  model_type: string;
  size: string;
  status: ImageGenStatus;
  has_thumbnail: boolean;
  file_size: number | null;
  width: number | null;
  height: number | null;
  duration_ms: number | null;
  created_at: string;
}

/** Full row returned by GET /image-generation/{id}. */
export interface ImageGenerationDetail extends ImageGenerationListItem {
  prompt: string;
  negative_prompt: string | null;
  quality: string | null;
  style: string | null;
  n: number;
  params: Record<string, unknown> | null;
  error_message: string | null;
  updated_at: string;
}

/** Body for POST /image-generation/. */
export interface ImageGenerationCreateRequest {
  model_config_id: number;
  prompt: string;
  negative_prompt?: string;
  size?: "1024x1024" | "1024x1792" | "1792x1024" | "512x512";
  n?: number;
  quality?: "standard" | "hd";
  style?: "vivid" | "natural";
  extra_params?: Record<string, unknown>;
  // M35: optional playbook id to inject style keywords into the prompt
  playbook_id?: number | null;
}

/** Body returned by POST /image-generation/ and /{id}/regenerate. */
export interface ImageGenerationCreateResponse {
  id: number;
  status: ImageGenStatus;
  batch_id: string | null;
  model_config_id: number;
  created_at: string;
}
